dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java

```diff
@@ -252,17 +252,20 @@ void addIndicatorValues( DataQueryParams params, Grid grid )
 
             List<List<DimensionItem>> dimensionItemPermutations = dataSourceParams.getDimensionItemPermutations();
 
+            Map<DimensionalItemId, DimensionalItemObject> itemMap = expressionService
+                .getIndicatorDimensionalItemMap( indicators );
+
             Map<String, List<DimensionItemObjectValue>> permutationDimensionItemValueMap = getPermutationDimensionItemValueMap(
-                dataSourceParams );
+                params, new ArrayList<>( itemMap.values() ) );
```

Comment:
I guess the task of the `ArrayList` wrapper here is to either make generics fit and/or avoid manipulation of the original collection. I think we should not copy for those reasons. Most of the time `Collections.unmodifiableList` can solve this without a copy.

Reply:
The task of the ArrayList wrapper is to return a List whereas Map.values() returns a Collection. Collections.unmodifiableList doesn't work here because that method requires a List argument, not a Collection.

My first thought was to change the getAggregatedDataValueMap to accept a Collection instead of a List, and I started making this change. However getAggregatedDataValueMap calls GridUtils.getGridIndexByDimensionItem which requires a List and steps through the list by index so it can return the index. So this just pushes the problem down into getAggregatedDataValueMap to convert the Collection to a List before calling GridUtils.getGridIndexByDimensionItem.

getAggregatedDataValueMap also passes the same Collection/List to a BaseDimensionalObject constructor which requires a List to set a List property, so getAggregatedDataValueMap would have to do the same conversion a second time for this purpose.

It seemed better to do the conversion only once, and not change the arguments in these other methods.